providers/mlx5/mlx5.c

```diff
@@ -388,44 +377,46 @@ int mlx5_destroy_psv(struct mlx5_psv *psv)
 	return ret;
 }
 
+#if defined(__i386__)
+static unsigned int get_mask(int startbit, int endbit)
+{
+	int i;
+	int result = 0;
+
+	for (i = startbit; i >= endbit; i--)
+		result |= (1 << i);
+
+	return result;
+}
+
+#define EXTRACT_BITS(val, start, end) (((val) & get_mask((start), (end))) >> (end))
+#endif
+
 static int mlx5_is_sandy_bridge(void)
 {
-	char line[128];
-	FILE *fd;
-	int rc = 0;
-	int cur_cpu_family = -1;
-	int cur_cpu_model = -1;
+#if defined(__i386__)
+	unsigned int ax, bx, cx, dx;
+	unsigned int ext_family_id, ext_model_id, family_id, model_id;
 
-	fd = fopen("/proc/cpuinfo", "r");
-	if (!fd)
+	if (!__get_cpuid(1, &ax, &bx, &cx, &dx))
 		return 0;
 
-	while (fgets(line, 128, fd)) {
-		int value;
-
-		if (!strncmp(line, "processor", 9)) {
-			cur_cpu_family = -1;
-			cur_cpu_model  = -1;
-		} else if (!strncmp(line, "cpu family", 10)) {
-			if ((cur_cpu_family < 0) && (!read_number_from_line(line, &value)))
-				cur_cpu_family = value;
-		} else if (!strncmp(line, "model", 5)) {
-			if ((cur_cpu_model < 0) && (!read_number_from_line(line, &value)))
-				cur_cpu_model = value;
-		}
+	ext_family_id = EXTRACT_BITS(ax, 27, 20);
+	ext_model_id = EXTRACT_BITS(ax, 19, 16);
+	family_id = EXTRACT_BITS(ax, 11, 8);
+	model_id = EXTRACT_BITS(ax, 7, 4);
 
-		if (cur_cpu_family == -1 || cur_cpu_model == -1)
-			continue;
+	if (family_id == 6 || family_id == 15)
+		model_id = (ext_model_id << 4) + model_id;
 
-		/* if this is a Sandy Bridge CPU */
-		if ((cur_cpu_family == 6) &&
-		    (cur_cpu_model == 0x2A || (cur_cpu_model == 0x2D) ))
-			rc = 1;
-		break;
-	}
+	if (family_id == 15)
```

Comment:
Why do we need those lines if you return 1 in line 417 only when family_id == 6 ? if we have here 15 the below line 413 may not end-up with 6, isn't it ?